KMS and tendermint can't talk to each other (secret connection)

I tried using the KMS as a remote signer:

- on the kms side I set:
```yaml
   [validator.connection]
    type = "tcp"
    addr = "127.0.0.1"
    port = 8080
    secret_key_path = "tests/support/secret_connection.key"
``` 
- on the tendermint side I used (after `tendermint init`):
`/tendermint node --priv_validator_laddr=127.0.0.1:8080 --proxy_app=kvstore`

This causes:
```
18:14:49 [INFO] KMS node ID: rawed255191nqmq3sj94emmzmnrg5lk702332scc60d65fuzfu80wfd95rugmpsvm6af9
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: DecodeError { description: "buffer underflow", stack: [] }', libcore/result.rs:1009:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```
on the KMS side; and
```
I[12116-11-12|19:14:48.783] Starting TCPVal                              module=privval impl=TCPVal
E[12116-11-12|19:14:49.067] OnStart                                      module=privval err="unmarshal to [32]uint8 failed after 21 bytes (Mismatched byte array length: Expected 32, got 20): 1441A9424F4E1841725F390526C18B7AD67423013B673C8FFE1C1469958FEA6C"
ERROR: Failed to create node: Error starting private validator client: unmarshal to [32]uint8 failed after 21 bytes (Mismatched byte array length: Expected 32, got 20): 1441A9424F4E1841725F390526C18B7AD67423013B673C8FFE1C1469958FEA6C
```
or sometimes sth like:
```
E[12116-11-12|19:28:20.953] OnStart                                      module=privval err="unmarshal to [32]uint8 failed after 1 bytes (insufficient bytes decoding []byte of length 125): 7D4FEACC6C3A4E86C4302D7CBE33C873A7A9B1A47133D406B886B20885EB3F26"
``` 
on the tendermint side. 
Looks like an encoding/decoding problem. I'm debugging what is going on here, update the issue accordingly, and provide a fix.

Looks like the KMS and tendermint talk to each other again via #91. Tested using above commands (we still need a proper integration test soon).
 
<details>
  <summary>Output (tendermint), click to expand (Starting TCPVal ✅)</summary>

```
./tendermint node --priv_validator_laddr=127.0.0.1:8080 --proxy_app=kvstore
I[12116-11-12|22:57:27.420] Starting multiAppConn                        module=proxy impl=multiAppConn
I[12116-11-12|22:57:27.420] Starting localClient                         module=abci-client connection=query impl=localClient
I[12116-11-12|22:57:27.420] Starting localClient                         module=abci-client connection=mempool impl=localClient
I[12116-11-12|22:57:27.420] Starting localClient                         module=abci-client connection=consensus impl=localClient
I[12116-11-12|22:57:27.420] ABCI Handshake App Info                      module=consensus height=0 hash= software-version=0.15.0 protocol-version=1
I[12116-11-12|22:57:27.420] ABCI Replay Blocks                           module=consensus appHeight=0 storeHeight=0 stateHeight=0
I[12116-11-12|22:57:27.421] Completed ABCI Handshake - Tendermint and App are synced module=consensus appHeight=0 appHash=
I[12116-11-12|22:57:27.421] Starting TCPVal                              module=privval impl=TCPVal
I[12116-11-12|22:57:29.984] This node is not a validator                 module=consensus addr=F88883B673FC69D7869CAB098DE3BAFC2FF76EB8 pubKey=PubKeyEd25519{348550057E9B13D922943165A3AB3B5D71EE58FC742CEF66F8216E1CF9E5EBBF}
I[12116-11-12|22:57:29.994] P2P Node ID                                  module=p2p ID=6767a5557dd34647e5b3b6a8e9d5453be2c41545 file=/Users/ismail/.tendermint/config/node_key.json
I[12116-11-12|22:57:29.994] Starting Node                                module=main impl=Node
I[12116-11-12|22:57:29.994] Starting EventBus                            module=events impl=EventBus
I[12116-11-12|22:57:29.994] Starting PubSub                              module=pubsub impl=PubSub
I[12116-11-12|22:57:29.995] Starting RPC HTTP server on tcp://0.0.0.0:26657 module=rpc-server 
I[12116-11-12|22:57:29.995] Starting P2P Switch                          module=p2p impl="P2P Switch"
I[12116-11-12|22:57:29.995] Starting MempoolReactor                      module=mempool impl=MempoolReactor
I[12116-11-12|22:57:29.995] Starting BlockchainReactor                   module=blockchain impl=BlockchainReactor
I[12116-11-12|22:57:29.995] Starting BlockPool                           module=blockchain impl=BlockPool
I[12116-11-12|22:57:29.995] Starting ConsensusReactor                    module=consensus impl=ConsensusReactor
I[12116-11-12|22:57:29.995] ConsensusReactor                             module=consensus fastSync=true
I[12116-11-12|22:57:29.995] Starting EvidenceReactor                     module=evidence impl=EvidenceReactor
I[12116-11-12|22:57:29.995] Starting PEXReactor                          module=p2p impl=PEXReactor
I[12116-11-12|22:57:29.995] Starting AddrBook                            module=p2p book=/Users/ismail/.tendermint/config/addrbook.json impl=AddrBook
I[12116-11-12|22:57:29.995] Starting IndexerService                      module=txindex impl=IndexerService
I[12116-11-12|22:57:29.995] Ensure peers                                 module=p2p numOutPeers=0 numInPeers=0 numDialing=0 numToDial=10
I[12116-11-12|22:57:29.995] No addresses to dial nor connected peers. Falling back to seeds module=p2p 
I[12116-11-12|22:57:29.995] Started node                                 module=main nodeInfo="{ProtocolVersion:{P2P:4 Block:7 App:1} ID_:6767a5557dd34647e5b3b6a8e9d5453be2c41545 ListenAddr:tcp://0.0.0.0:26656 Network:test-chain-Qriebw Version:0.26.0 Channels:4020212223303800 Moniker:Ismails-MacBook-Pro.local Other:{TxIndex:on RPCAddress:tcp://0.0.0.0:26657}}"
E[12116-11-12|22:57:29.995] Couldn't connect to any seeds                module=p2p 
```
</details>

Output KMS:
```
 target/debug/tmkms run -c tmkms.toml
21:57:24 [INFO] tmkms 0.0.1 starting up...
21:57:24 [INFO] [keyring:dalek:gaia-8000] added validator key cosmosvalconspub1xjz4qpt7nvfajg55x9j682emt4c7uk8uwskw7ehcy9hpe709awlsdf2p7t
21:57:24 [INFO] KMS node ID: rawed255191nqmq3sj94emmzmnrg5lk702332scc60d65fuzfu80wfd95rugmpsvm6af9
21:57:24 [ERROR] [gaia-rpc://127.0.0.1:8080] I/O error: Connection refused (os error 61)
21:57:29 [INFO] KMS node ID: rawed255191nqmq3sj94emmzmnrg5lk702332scc60d65fuzfu80wfd95rugmpsvm6af9
21:57:29 [INFO] [gaia-9000@gaia-rpc://127.0.0.1:8080] connected to validator successfully

```

--- a/tendermint-rs/src/amino_types/secret_connection.rs
+++ b/tendermint-rs/src/amino_types/secret_connection.rs
@@ -2,7 +2,7 @@
 
 #[derive(Clone, PartialEq, Message)]
 pub struct AuthSigMessage {
-    #[prost(bytes, tag = "1")]
+    #[prost(bytes, tag = "1", amino_name = "tendermint/PubKeyEd25519")]
     pub key: Vec<u8>,
     #[prost(bytes, tag = "2")]
     pub sig: Vec<u8>,

--- a/tendermint-rs/src/secret_connection/mod.rs
+++ b/tendermint-rs/src/secret_connection/mod.rs
@@ -15,15 +15,12 @@ use self::{
 };
 use byteorder::{ByteOrder, LE};
 use bytes::BufMut;
-use prost::{
-    encoding::{bytes::merge, encode_varint, WireType},
-    Message,
-};
+use prost::{encoding::encode_varint, Message};
 use signatory::{ed25519, Signature, Signer};
 use signatory_dalek::Ed25519Verifier;
 use std::{
     cmp,
-    io::{self, Cursor, Read, Write},
+    io::{self, Read, Write},
     marker::{Send, Sync},
 };
 
@@ -285,27 +282,29 @@ fn share_eph_pubkey<IoHandler: Read + Write + Send + Sync>(
 
     let mut buf = vec![0; 0];
     let local_eph_pubkey_vec = &local_eph_pubkey.to_vec();
-    // Note: this is not regular protobuf encoding but raw length prefixed amino encoding:
-    encode_varint(local_eph_pubkey_vec.len() as u64, &mut buf);
-    buf.put_slice(local_eph_pubkey_vec);
+    // Note: this is not regular protobuf encoding but raw length prefixed amino encoding;
+    // amino prefixes with the total length, and the raw bytes array's length, too:
+    encode_varint((local_eph_pubkey_vec.len() + 1) as u64, &mut buf); // 33
+    encode_varint(local_eph_pubkey_vec.len() as u64, &mut buf); // 32
+    buf.put_slice(local_eph_pubkey_vec); // raw bytes
+
+    // TODO(ismail): we probably do *not* need the double length delimiting here or in tendermint)
     // this is the sending part of:
     // https://github.com/tendermint/tendermint/blob/013b9cef642f875634c614019ab13b17570778ad/p2p/conn/secret_connection.go#L208-L238
     handler.write_all(&buf)?;
 
-    let mut buf = vec![0; 33];
+    let mut buf = vec![0; 34];
     handler.read_exact(&mut buf)?;
 
-    let mut amino_buf = Cursor::new(buf);
     // this is the receiving part of:
     // https://github.com/tendermint/tendermint/blob/013b9cef642f875634c614019ab13b17570778ad/p2p/conn/secret_connection.go#L208-L238
-    let mut remote_eph_pubkey = vec![];
-    merge(
-        WireType::LengthDelimited,
-        &mut remote_eph_pubkey,
-        &mut amino_buf,
-    ).unwrap();
     let mut remote_eph_pubkey_fixed: [u8; 32] = Default::default();
-    remote_eph_pubkey_fixed.copy_from_slice(&remote_eph_pubkey[..32]);
+    if buf[0] != 33 || buf[1] != 32 {
+        return Err(Error::Protocol);
+    }
+    // after total length (33) and byte length (32), we expect the raw bytes
+    // of the pub key:
+    remote_eph_pubkey_fixed.copy_from_slice(&buf[2..34]);
 
     Ok(remote_eph_pubkey_fixed)
 }
@@ -339,15 +338,14 @@ fn share_auth_signature<IoHandler: Read + Write + Send + Sync>(
         sig: signature.into_bytes().to_vec(),
     };
     let mut buf: Vec<u8> = vec![];
-    amsg.encode(&mut buf)?;
-
+    amsg.encode_length_delimited(&mut buf)?;
     sc.write_all(&buf)?;
 
-    let mut rbuf = vec![0; 100]; // 100 = 32 + 64 + (amino overhead)
+    let mut rbuf = vec![0; 106]; // 100 = 32 + 64 + (amino overhead = 2 fields + 2 lengths + 4 prefix bytes + total length)
     sc.read_exact(&mut rbuf)?;
 
     // TODO: proper error handling:
-    Ok(AuthSigMessage::decode(&rbuf)?)
+    Ok(AuthSigMessage::decode_length_delimited(&rbuf)?)
 }
 
 #[cfg(tests)]

